File: app/src/main/java/torille/fi/lurkforreddit/data/RedditRepository.java
package torille.fi.lurkforreddit.data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.Pair;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import torille.fi.lurkforreddit.data.models.view.Comment;
import torille.fi.lurkforreddit.data.models.view.Post;
import torille.fi.lurkforreddit.data.models.view.SearchResult;
import torille.fi.lurkforreddit.data.models.view.Subreddit;
import torille.fi.lurkforreddit.di.scope.RedditScope;

/**
 * Created by eva on 23.4.2017.
 */
@RedditScope
public class RedditRepository implements RedditDataSource {

    private final RedditDataSource mRedditRemoteApi;

    @Nullable
    private List<Subreddit> mCachedSubreddits;

    @Inject
    public RedditRepository(@Remote RedditDataSource redditApi) {
        mRedditRemoteApi = redditApi;
    }

    @Override
    public Observable<List<Subreddit>> getSubreddits() {
        if (mCachedSubreddits == null) {
            return mRedditRemoteApi.getSubreddits()
                    .flatMap(new Function<List<Subreddit>, Observable<List<Subreddit>>>() {
                        @Override
                        public Observable<List<Subreddit>> apply(@io.reactivex.annotations.NonNull List<Subreddit> subreddits) throws Exception {
                            return Observable.fromArray(subreddits)
                                    .doOnNext(new Consumer<List<Subreddit>>() {
                                        @Override
                                        public void accept(@io.reactivex.annotations.NonNull List<Subreddit> subreddits) throws Exception {
                                            mCachedSubreddits = subreddits;
                                        }
                                    });
                        }
                    });
        } else {
            return  Observable.fromArray(mCachedSubreddits);
        }
    }

    @Override
    public Observable<Pair<String, List<Post>>> getSubredditPosts(@NonNull String subredditUrl) {
        return mRedditRemoteApi.getSubredditPosts(subredditUrl);
    }

    @Override
    public Observable<Pair<String, List<Post>>> getMoreSubredditPosts(@NonNull String subredditUrl, @NonNull String nextpageId) {
        return mRedditRemoteApi.getMoreSubredditPosts(subredditUrl, nextpageId);
    }

    @Override
    public void refreshData() {
        mCachedSubreddits = null;
    }

    @Override
    public Observable<List<Comment>> getCommentsForPost(@NonNull String permaLinkUrl) {
        return mRedditRemoteApi.getCommentsForPost(permaLinkUrl);
    }

    @Override
    public Observable<List<Comment>> getMoreCommentsForPostAt(@NonNull List<String> childCommentIds,
                                                              @NonNull String linkId,
                                                              int commentLevel) {
        return mRedditRemoteApi.getMoreCommentsForPostAt(childCommentIds, linkId, commentLevel);
    }

    @Override
    public Observable<Pair<String, List<SearchResult>>> getSearchResults(@NonNull String query) {
        return mRedditRemoteApi.getSearchResults(query);
    }

    @Override
    public Observable<Pair<String, List<SearchResult>>> getMoreSearchResults(@NonNull String query, @NonNull String afterId) {
        return mRedditRemoteApi.getMoreSearchResults(query, afterId);
    }
}

File: app/src/main/java/torille/fi/lurkforreddit/data/RedditRepository.kt
package torille.fi.lurkforreddit.data

import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import torille.fi.lurkforreddit.data.models.view.*
import torille.fi.lurkforreddit.di.scope.Local
import torille.fi.lurkforreddit.di.scope.Remote
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RedditRepository @Inject
constructor(@Remote private val redditRemoteApi: RedditDataSource, @Local private val redditLocalApi: RedditDataSource) :
    RedditDataSource {

    override fun saveSubreddits(subreddits: List<Subreddit>) {
        return redditLocalApi.saveSubreddits(subreddits)
    }

    override fun getSubreddits(): Flowable<List<Subreddit>> {
        return redditLocalApi
            .getSubreddits()
            .map {
                if (it.isNotEmpty()) {
                    Flowable.fromArray(it)
                } else {
                    redditRemoteApi.getSubreddits()
                        .observeOn(Schedulers.io())
                        .doOnNext {
                            redditLocalApi.saveSubreddits(it)
                        }
                }
            }.flatMap { it }

    }

    override fun saveSubreddit(subreddit: Subreddit) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getSubredditPosts(subredditUrl: String): Observable<kotlin.Pair<String, List<Post>>> {
        return redditRemoteApi.getSubredditPosts(subredditUrl)
    }

    override fun getMoreSubredditPosts(
        subredditUrl: String,
        nextpageId: String
    ): Observable<kotlin.Pair<String, List<Post>>> {
        return redditRemoteApi.getMoreSubredditPosts(subredditUrl, nextpageId)
    }

    override fun refreshData() {
        redditLocalApi.refreshData()
    }

    override fun getCommentsForPost(
        permaLinkUrl: String,
        isSingleCommentThread: Boolean
    ): Observable<PostAndComments> {
        return redditRemoteApi.getCommentsForPost(permaLinkUrl, isSingleCommentThread)
    }

    override fun getMoreCommentsForPostAt(
        childCommentIds: List<String>,
        linkId: String,
        commentLevel: Int
    ): Observable<List<Comment>> {
        return redditRemoteApi.getMoreCommentsForPostAt(childCommentIds, linkId, commentLevel)
    }

    override fun getSearchResults(query: String): Flowable<kotlin.Pair<String, List<SearchResult>>> {
        return redditRemoteApi.getSearchResults(query)
    }

    override fun getMoreSearchResults(
        query: String,
        after: String
    ): Flowable<kotlin.Pair<String, List<SearchResult>>> {
        return redditRemoteApi.getMoreSearchResults(query, after)
    }
}
